External/Plugins/ASCompletion/Completion/ASComplete.cs

```diff
@@ -3777,6 +3829,100 @@ static public ASResult GetExpressionType(ScintillaControl sci, int position, boo
             }
         }
 
+        static private MemberList GetTypeParameters(MemberModel model)
+        {
+            MemberList retVal = null;
+            string template = model.Template;
+            if (template != null && template.StartsWith("<"))
+            {
+                var sb = new StringBuilder();
+                int groupCount = 0;
+                bool inConstraint = false;
+                MemberModel genType = null;
+                for (int i = 1, count = template.Length - 1; i < count; i++)
+                {
+                    char c = template[i];
+                    if (!inConstraint)
+                    {
+                        if (c == ':' || c == ',')
+                        {
+                            genType = new MemberModel();
+                            genType.Name = sb.ToString();
+                            genType.Type = sb.ToString();
+                            genType.Flags = FlagType.TypeDef;
+                            inConstraint = c == ':';
+                            if (retVal == null) retVal = new MemberList();
+                            retVal.Add(genType);
+                            sb.Length = 0;
+
+                            continue;
+                        }
+                        else if (char.IsWhiteSpace(c)) continue;
+                        sb.Append(c);
+                    }
+                    else
+                    {
+                        if (c == ',')
+                        {
+                            if (groupCount == 0)
+                            {
+                                genType.Type += ":" + sb.ToString();
+                                genType = null;
+                                inConstraint = false;
+                                sb.Length = 0;
+                                continue;
+                            }
+                        }
+                        else if ("({[<".IndexOf(c) > -1)
+                            groupCount++;
+                        else if (")}]>".IndexOf(c) > -1)
+                            groupCount--;
+                        sb.Append(c);
+                    }
+                }
+                if (sb.Length > 0)
+                {
+                    if (retVal == null) retVal = new MemberList();
+                    if (!inConstraint)
+                        retVal.Add(new MemberModel { Name = sb.ToString(), Type = sb.ToString(), Flags = FlagType.TypeDef });
+                    else
+                        genType.Type += ":" + sb.ToString();
+                }
+            }
+
+            return retVal;
+        }
+
+        static private MemberList GetVisibleElements()
+        {
+            MemberList known = new MemberList();
+            known.Add(ASContext.Context.GetVisibleExternalElements());
```

Comment:
Since it uses AddRange it shouldn't be very noticeable, but will change it. It's also a readability thing.